### PR TITLE
WIP: Refactor the calendar plugin and fix the navigation

### LINKED ISSUE #6558: Datepicker Redesign to Reduce User Errors
Last week, we had usability participants try to enter their date of birth via the datepicker. People had many problems that I will explain in other issues, but for now, would like to understand if this particular problem is a bug, because it looks like one. 

Participant opens the calendar, selects year and month, clicks GO, then clicks a day, and it reverts back to the current date.  On Chrome, it does NOT revert to the current date. Participant is on Firefox v35.0.1. See the video here (participant voice is disguised): http://www.youtube.com/watch?v=38FBlQud3wc

I can't replicate this on the latest version of firefox on a Mac. 

Working version is here. http://neoinsight.com/test/wetform1/exform2-en.html
I don't think it's a code problem because it worked for other people. 

<div class="form-group">
  <label for="bdate1"><span class="field-name">When were you born?</span>
  <span class="datepicker-format"> (YYYY-MM-DD)</span> </label>
<p> Please use your real date of birth - just switch the month to one month different from your own.</p>
<input class="form-control" id="bdate1" name="bdate1" type="date" data-rule-dateiso="true">
</div>

### PR DESCRIPTION
**Pull request by LaurentGoderre**
Fixes #6558, #6823, #6683, #6689, #6599, #6441, #6744, #7047, #6697

### CODE REVIEW

**Comment by andre-gh** [comment]
Can this pull request be closed, has it gone stale?

**Comment by LaurentGoderre** [comment]
No, we still need this work, I just need to fin the time to finish it

File: src/polyfills/datepicker/test.js [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+				};
+				beforeFactory()();
+			} );
+			after( function() {
+				$elm.next().click();
+				defaultAfter();
+			} );
+
+			it( "should not have populated the form field", function() {
+				expect( $elm.val() ).to.equal( "" );
```
Comment (LaurentGoderre) [inline src/polyfills/datepicker/test.js]: Unit test that proves #6823 is fixed

**Comment by LaurentGoderre** [comment]
Need to test #6441

File: src/polyfills/datepicker/test.js [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+			} );
+			after( defaultAfter );
+
+			it( "should have populated the from field with the selected date", function() {
+				expect( $elm.val() ).to.equal( "2014-03-06" );
+			} );
+
+			it( "should have triggered the change event on the form field", function() {
+				expect( spy.calledWith( "change" ) ).to.equal( true );
+				expect( spy.calledOn( $elm ) ).to.equal( true );
```
Comment (LaurentGoderre) [inline src/polyfills/datepicker/test.js]: Test top prove #6683 is solved

File: src/polyfills/datepicker/test.js [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+				defaultAfter();
+			} );
+			it( "should have set the calendar to the same month as the field's date", function() {
+				var settings = $calendar.get( 0 ).firstChild.lib;
+				expect( settings.year ).to.equal( 2014 );
+				expect( settings.month ).to.equal( 7 );
+			} );
+
+			it( "should have highlighted the selected date", function() {
+				expect( $calendar.find( ".cal-index-7 > a" ).attr( "aria-selected" ) ).to.equal( "true" );
```
Comment (LaurentGoderre) [inline src/polyfills/datepicker/test.js]: Test to prove #6744 was fixed

File: src/polyfills/datepicker/test.js [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+						"</span>" +
+					"</strong>" +
+				"</label>";
+
+			before( beforeFactory( null, label ) );
+			after( defaultAfter );
+
+			it( "should have stored only the field name in the label", function() {
+				var state = $elm.get( 0 ).state;
+				expect( state.labelText ).to.equal( "Appointment Date" );
```
Comment (LaurentGoderre) [inline src/polyfills/datepicker/test.js]: Test to prove #7047 was fixed

**Comment by LaurentGoderre** [comment]
@masterbee can we merge this? I have not yet done the unit test for the event calendar and as I do I am sure I will discover some minor bug but I don't think it's enough to hold off all the fixes to the date picker this brings

/cc @neoinsight I this is finally ready to go in!

**Comment by neoinsight** [comment]
@LaurentGoderre Congratulations! 
@masterbee @pcwsmith Once it is live on a demo page, we will schedule it into a testing round, including with screenreader users.

**Comment by LaurentGoderre** [comment]
9 months in the making......assuming it gets merged soon ;)

**Comment by LaurentGoderre** [comment]
@andre-gh can we merge this?

**Comment by LaurentGoderre** [comment]
Thanks! Feels so great to close all these issues. @neoinsight this is ready! Hope testing goes well!

**Comment by Rantanpl4n18** [comment]
Problem still there for issue #6441. I just downloaded the 4.0.20 production files and the calendar still have the same problem.

**Comment by LaurentGoderre** [comment]
Can you provide an example?

**Comment by Rantanpl4n18** [comment]
Here the look before and after triggering the calendar with 4.0.20 :
nssp cal_1
nssp cal_2

**Comment by LaurentGoderre** [comment]
I don't think it's the same issues. I would suspect it's more related to this new issues #7395

**Comment by Rantanpl4n18** [comment]
Well in #6441, the datepicker is limited Inside the tab instead of going over. That is exactly what's happening here. Even if the focus stay on the calendar or go at the bottom of the document, the calendar is still stucked Inside tha tab. 

Am I missing something?

Thanks

**Comment by LaurentGoderre** [comment]
Oh you're right, but that is not the latest version of the calendar, that is still the old one.

**Comment by LaurentGoderre** [comment]
Aaaah, there hasn't been a WET release since this was merged. This will only be included in v4.0.21. Not sure when that will be.

**Comment by Rantanpl4n18** [comment]
Oh :(
Is the BETA version available or something like that?
Thanks

**Comment by LaurentGoderre** [comment]
You can always download the latest version here: https://github.com/wet-boew/wet-boew-cdn/archive/v4.0-dist.zip

**Comment by LaurentGoderre** [comment]
Or if you are using bower to include it, you can use

```
"dependencies": {
    "wet-boew": "wet-boew/wet-boew-cdn#v4.0-dist",
  }
```

**Comment by Rantanpl4n18** [comment]
Allright, thanks. Got the latest version, it works, the calendar goes over the tab. There is one remaning problem: the calendar background is transparent. Maybe one of our style override the calendar CSS. I'll look forward to fix this, :)
nssp cal_3

**Comment by LaurentGoderre** [comment]
Did you get the updated CSS as well? What theme are you using?

**Comment by Rantanpl4n18** [comment]
Sorry for late reply, I wasnt at work.

Mmm I only downloaded this : https://github.com/wet-boew/wet-boew-cdn/archive/v4.0-dist.zip

Is there anything else?

For the theme, I'm not sure. We have some guys in our team who created a "base project", so everything is already included. Might be the "Intranet" theme.

**Comment by LaurentGoderre** [comment]
The theme contains the styles for the Calendar (so that it can customize it's look to fit with the theme). You need to find out what theme you are using (GCWeb, GC Web Usabilty or Intranet) and I can point you to the other package you need to download

**Comment by Rantanpl4n18** [comment]
We use the intranet theme : 
http://wet-boew.github.io/themes-dist/theme-gc-intranet/content-fr.html

**Comment by LaurentGoderre** [comment]
You need this zip as well https://github.com/wet-boew/themes-cdn/archive/theme-gc-intranet.zip

**Comment by Rantanpl4n18** [comment]
Allright, now the calendar works. Background is not transparent anymore and the bug is fixed. :)
Only remaining problem : all glyphicons don't work anymore... Maybe I will revert all changes and update only the calendar file, might be easier...
glyph missing

Thanks

**Comment by LaurentGoderre** [comment]
You can't just update the calendar file because it depends on a core functionality. Did you make sure that the folder structure respects the new structure?

```
wet-boew/
theme-gc-intranet/
```

**Comment by Rantanpl4n18** [comment]
Mmm I don't see any change in the structure, it's the same. The Glyphicons files are in the "fonts" folder
wet4 structure

**Comment by LaurentGoderre** [comment]
A few version ago, the structure was changed to better accommodate CDNs, WET core and the theme need to be side-by-side in the structure mentioned above.

**Comment by Rantanpl4n18** [comment]
Ok got it. That works. But I have the issue #7395 that wasn't present before. So I have to choose which bug I want : the new one or the old one :P

I'll try to fix the focus.

Thanks for your help, really appreciated!

**Comment by LaurentGoderre** [comment]
The other bug should be fixed shortly.

**Comment by Rantanpl4n18** [comment]
Ok thanks.

**Comment by LaurentGoderre** [comment]
@Rantanpl4n18 if you redownload this package, it should work https://github.com/wet-boew/wet-boew-cdn/archive/v4.0-dist.zip

**Comment by Rantanpl4n18** [comment]
Thanks for the update!

**Pull request merged**